### PR TITLE
adding Product.Iterate

### PR DESCRIPTION
**Pull request by luthermonson**
Saw the merging of ListWithPagination and thought I'd try and add my iterators to the mix as well. I made some of these to make my life easier and thought it'd be better in the package. Looking for approval from maintainers that the idea is sound as I'd like to add more.

Usage:
```
err := client.Product.IterateAll(func(product goshopify.Product) error {
    fmt.Println(product.Title)
})
```

### CODE REVIEW

**Comment by luthermonson** [comment]
I will add tests once feedback is given.

**Comment by epelc** [comment]
@luthermonson I think if we are going to include helpers for mass iteration they should include context support so you can cancel them since these could easily take a bit to run.

Can you consider adding a context and cancelation support? Also might be interesting to investigate bringing the new pagination support to all other resources and standardizing the bulk of this helper code so we can use on all endpoints. I reviewed and helped cleanup the original pagination list PR and might need order pagination in a bit so might take this on.

Anyways I think for this to merge it should take a context and handle the cancelation. LMK if you need assistance with implementing this. Looks good otherwise.

**Comment by epelc** [comment]
Also another major point is it should switch to the new ` ListWithPagination` since it's merged now and Shopify is ending support for the older skip+limit pagination very soon(I believe in a month or two).

File: product.go [hunk 1]
```diff
@@ -232,6 +237,45 @@ func (s *ProductServiceOp) Delete(productID int64) error {
...
+			break
+		}
+
+		for _, product := range products {
+			if err := f(product); err != nil {
+				return err
+			}
+
+			sinceId = product.ID
+		}
```
Comment (ysurtayev-bold) [inline product.go]: adding the check ```if len(product)<options.Limit { break }```  here will allow to save one call to Shopify.
I.e. if we get less records than we requested - there is no next page.

**Pull request closed without merging**